File: python/hard/largest_range/largest_range.py
# O(n) time | O(n) space, where n is the number of elements in the array
def largestRange(array):
    # Initialize variables to store the best range and the length of the longest range found so far
    bestRange = []
    longestLength = 0
    # Create a dictionary to keep track of the numbers that have been visited
    nums = {}

    # Populate the dictionary, setting all numbers to True, indicating they have not been visited yet
    for num in array:
        nums[num] = True

    # Iterate through each number in the array
    for num in array:
        # If the number has already been visited, skip it
        if not nums[num]:
            continue

        # Mark the number as visited
        nums[num] = False
        # Initialize the length of the current range to 1 (the current number itself)
        currentLength = 1
        # Set up pointers to check numbers to the left and right of the current number
        left = num - 1
        right = num + 1

        # Expand the range to the left
        while left in nums:
            # Mark the left number as visited
            nums[left] = False
            # Increase the length of the current range
            currentLength += 1
            # Move the left pointer further left
            left -= 1

        # Expand the range to the right
        while right in nums:
            # Mark the right number as visited
            nums[right] = False
            # Increase the length of the current range
            currentLength += 1
            # Move the right pointer further right
            right += 1

        # If the current range is longer than the longest range found so far, update the best range
        if currentLength > longestLength:
            longestLength = currentLength
            # Store the range as [left + 1, right - 1], since left and right are one step beyond the actual range
            bestRange = [left + 1, right - 1]

    # Return the best (largest) range found
    return bestRange

File: java/hard/largest_range/LargestRange.java
package main.java.com.example.largestrange;

import java.util.HashMap;
import java.util.Map;

public class LargestRange {
	// O(n) time | O(n) space, where n is the number of elements in the array
    public static int[] largestRange(int[] array) {
        // Initialize variables to store the best range and the length of the longest range found so far
        int[] bestRange = new int[2];
        int longestLength = 0;

        // Create a HashMap to keep track of the numbers that have been visited
        Map<Integer, Boolean> nums = new HashMap<>();

        // Populate the HashMap, setting all numbers to true, indicating they have not been visited yet
        for (int num : array) {
            nums.put(num, true);
        }

        // Iterate through each number in the array
        for (int num : array) {
            // If the number has already been visited, skip it
            if (!nums.get(num)) {
                continue;
            }

            // Mark the number as visited
            nums.put(num, false);
            // Initialize the length of the current range to 1 (the current number itself)
            int currentLength = 1;
            // Set up pointers to check numbers to the left and right of the current number
            int left = num - 1;
            int right = num + 1;

            // Expand the range to the left
            while (nums.containsKey(left)) {
                // Mark the left number as visited
                nums.put(left, false);
                // Increase the length of the current range
                currentLength++;
                // Move the left pointer further left
                left--;
            }

            // Expand the range to the right
            while (nums.containsKey(right)) {
                // Mark the right number as visited
                nums.put(right, false);
                // Increase the length of the current range
                currentLength++;
                // Move the right pointer further right
                right++;
            }

            // If the current range is longer than the longest range found so far, update the best range
            if (currentLength > longestLength) {
                longestLength = currentLength;
                // Store the range as [left + 1, right - 1], since left and right are one step beyond the actual range
                bestRange[0] = left + 1;
                bestRange[1] = right - 1;
            }
        }

        // Return the best (largest) range found
        return bestRange;
    }
}